### ISSUE TITLE
Cargo install --git fails with repo using sha256 object format

### ISSUE THREAD

**Issue by Joshix-1** [rust-lang/cargo#14942]
### Problem

`cargo install --git https://codeberg.org/Joshix/rps` fails

### Steps

```
$ cargo install --git https://codeberg.org/Joshix/rps
    Updating git repository `https://codeberg.org/Joshix/rps`
error: failed to fetch into: /home/josh/.cargo/git/db/rps-7d405a448b2c5f11

Caused by:
  unknown remote object format 'sha256'; class=Invalid (3)
```

```
$ cargo +nightly -Zgitoxide install --git https://codeberg.org/Joshix/rps
    Updating git repository `https://codeberg.org/Joshix/rps`
error: failed to clone into: /home/josh/.cargo/git/db/rps-257013808e5f3c48

Caused by:
  failed to fill whole buffer
```

```
$ env CARGO_NET_GIT_FETCH_WITH_CLI=true cargo install --git https://codeberg.org/Joshix/rps
    Updating git repository `https://codeberg.org/Joshix/rps`
fatal: mismatched algorithms: client sha1; server sha256
error: failed to clone into: /home/josh/.cargo/git/db/rps-7d405a448b2c5f11

Caused by:
  process didn't exit successfully: `git fetch --force --update-head-ok 'https://codeberg.org/Joshix/rps' '+HEAD:refs/remotes/origin/HEAD'` (exit status: 128)
```

### Possible Solution(s)

Support the sha256 object format

### Version

```text
cargo 1.83.0 (5ffbef321 2024-10-29)
release: 1.83.0
commit-hash: 5ffbef3211a8c378857905775a15c5b32a174d3b
commit-date: 2024-10-29
host: x86_64-unknown-linux-gnu
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.9.0-DEV (sys:0.4.74+curl-8.9.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Arch Linux Rolling Release [64-bit]
```

**Comment by weihanglo** [comment]
Kinda blocked on <https://github.com/rust-lang/git2-rs/issues/1090> and <https://github.com/GitoxideLabs/gitoxide/issues/281>

Still an upstream issue, though `git2` is within cargo team's reach.

**Comment by weihanglo** [comment]
cc @Byron
no pressure :)

**Comment by Byron** [comment]
Thanks for the hint 😅!
I have added it to the list of requirements for Cargo, and once `git2` adds the support for it that would pressure `gitoxide` to add such support as well.
On the bright side, support for it is already built-in to all of the code, but to assure basic functioning a lot of strategic tests would have to be written.

**Comment by Joshix-1** [comment]
Why doesn't cargo run `git clone` when `CARGO_NET_GIT_FETCH_WITH_CLI=true` is set? That could probably fix it with that setting.

**Comment by weihanglo** [comment]
Cargo internally maintains git datatbases from different repository. They are initialized and 
 fetched in separate steps. We don't use `git clone` directly because 1) it has less customization points and 2) we don't want to assume a specific CLI tool exist on user's machine.

To make sha256 objectFormat just works for git CLI, we may need

* When Cargo initializes a git database, set the `extension.objectFormat` to `sha256` (and `repositoryformatversion` to `1` maybe?). I am not particularly familiar with [the migration plan](https://git-scm.com/docs/hash-function-transition) and how to set up these configs properly.
  * It might be possible to check [`object-format`](https://git-scm.com/docs/protocol-capabilities#_object_format) capability during git negotiation, and set the necessary configurations.
* Internally Cargo checkouts source code to another directory when in use, this "local checkout" is facilitated by libgit2 now. To make it work, libgit2 must know how to do sha256 object format. The support is still [in question](https://github.com/libgit2/libgit2/blob/4dcdb64c6844d76776745cdc25071a72c1af84d6/CMakeLists.txt#L19), and could be a hard blocker.

Git CLI support was added for helping some weird networking issue on user side. From my understanding it is not a tool Cargo pursues for all its shiny new features, like [lacking shallow clone support from git CLI integration](https://github.com/rust-lang/cargo/issues/14956). We normally libgit2 as baseline, and nowadays gitoxide is our pioneer :)

**Comment by msrd0** [comment]
> Git CLI support was added for helping some weird networking issue on user side

@weihanglo Git CLI is necessary whenever I want to download private git repositories. Cargo tries to get an ssh key from the ssh-agent for the username `git` (which fails obviously) while the git cli just uses my default ssh key which works. Not sure if I'd classify that as "some weird networking issue".

**Comment by weihanglo** [comment]
@msrd0 
Guess you're looking for https://github.com/rust-lang/cargo/issues/2078

**Comment by msrd0** [comment]
@weihanglo Well I don't have entries related to git at all in my ssh config so I'm not sure that's the correct issue. (Tbh I always thought that cargo just didn't properly support authenticating with ssh, it was only after I read your statement earlier today, so I never searched for any open issues)

**Comment by A248** [comment]
The lack of sha256 object support also seems to affect other areas.

For example, Cargo's detection of .gitignore. Ignored files are included in `cargo publish`, and when I debug with verbosity enabled, I see this line: `warning: no (git) VCS found for /home/myuser/...`.

Should I open a new issue or just use this one?

**Comment by weihanglo** [comment]
@A248 I changed the title to be more general. For sure that it is affecting beyond `cargo install`. It would be appreciated if people could help identify which Cargo features are affected by the lack of sha256 support for Git. The git2-rs support is slowloy progressing: <https://github.com/rust-lang/git2-rs/pull/1201>.

**Comment by weihanglo** [comment]
Update: Some recent works in upstream

* https://github.com/rust-lang/git2-rs/pull/1201
* https://github.com/rust-lang/git2-rs/pull/1202
* https://github.com/libgit2/libgit2/pull/7179
* https://github.com/libgit2/libgit2/pull/7183
* https://github.com/libgit2/libgit2/pull/7185

Once they've merged, we might be able to start adding `-Zgit=sha256`.

However, there is an issue that we must turn on the conditional compilation flag `EXPERIMENTAL_SHA256` in libgit2 always, which is a bit unfortunate. We can only detect SHA256 support at runtime. For example, before fetching, check if a `remote.object_format()` is SHA256 and bail if `-Zgit=sha256` is not enabled.

Yet the above workaround has caveat: Cargo relies on git not only for fetching dependencies, but also source file listing and package dirtiness check (some code paths are on gitoxide). We might want separate sub unstable features to support different cases. We shouldn't implicitly support `cargo package` working in SHA256 repository without nightly flag.

**Comment by A248** [comment]
Just curious about your last sentence. Why would sha256 support be put behind nightly/unstable features? It's not like Sha256 is a feature developed _by Cargo_. More like fixing a bug / updating to the latest version of an integration.

Sha256 in Git is considered stable as far as I know, so this issue is more about getting Cargo up-to-speed on standards. Nothing special about it otherwise.

**Comment by weihanglo** [comment]
> Just curious about your last sentence. Why would sha256 support be put behind nightly/unstable features? It's not like Sha256 is a feature developed _by Cargo_. More like fixing a bug / updating to the latest version of an integration.


I mostly agree that this can be viewed as a bugfix, but it is not a small or isolated one, especially given the number of affected code paths and the current state of upstream support.

In Cargo, changes of this size usually start behind a nightly feature so we can gather feedback before stabilization. This also helps coordinate work across multiple parties and allows us to land the work incrementally rather than all at once.

And for the state of upstream support, the relevant dependencies (both libgit2 and gitoxide) have not yet fully stabilized their SHA-256 support and still consider it experimental. As noted in my earlier comment <https://github.com/rust-lang/cargo/issues/14942#issuecomment-3711859556>, I have already run into some issues in libgit2, which is exactly the kind of thing we usually want to catch during a nightly phase.

**Comment by A248** [comment]
> It would be appreciated if people could help identify which Cargo features are affected by the lack of sha256 support for Git.

Another affected feature is the commit hash published with crates.io package metadata. This hash is typically used to verify that crates.io package source code matches the repository source code.

----------------------------------------------------------------------------------------------------------------------------------------------------------------------

That brings the affected features to the following list, so far:

* Running `cargo install --git <URL>`
* Using .gitignore to exclude files from published package
* The commit hash included in crates.io package metadata

(I can update this comment and add more list items as people identify more affected features)

**Comment by weihanglo** [comment]
Thanks!

> The commit hash included in crates.io package metadata

Which one is it? The [`.cargo_vcs_info.json` file](https://doc.rust-lang.org/nightly/cargo/commands/cargo-package.html#cargo_vcs_infojson-format)?

**Comment by A248** [comment]
Yes, I'm pretty sure. I encountered this in context of lib.rs reporting that my crates have no git commmit hash, [for example](https://lib.rs/~A248/dash). I wasn't aware of the specific file, but searching for `.cargo_vcs_info.json` in the [lib.rs repository](https://gitlab.com/lib.rs/main) gives plenty of results.

**Comment by bk2204** [comment]
I'll point out that Git 3.0 will likely be out later this year (probably around September) and SHA-256 repositories and reftable (see #15184) will be the default.  That will happen whether or not libgit2 or the `git2` crate have support for those features.

That does mean that if Cargo continues to depend on a version of libgit2 without those features, it will not work out of the box with repositories created by `git init` or `git clone` once Git 3.0 comes out.  I suggest that work be undertaken promptly to get Cargo fully functional.  (You can compile Git with `WITH_BREAKING_CHANGES=1` to test out the Git 3.0 features.)

**Comment by weihanglo** [comment]
Thanks for the info. We are aware of the timeline of Git 3.0, hence the changes in <https://github.com/rust-lang/git2-rs/issues/1090> (and relevant upstream libgit2 contributions). And I have a local branch for SHA256 support in Cargo with the draft git2-rs PR. Currently waiting for libgit2 to get a new release out so

* they can get feedback from git2-rs users, and
* we can start the integration sooner.

See also <https://github.com/libgit2/libgit2/discussions/7149#discussioncomment-16416597> (just pinged the libgit2 maintainer)

**Comment by Byron** [comment]
`gitoxide` will also be working hard to make it happen.

**Comment by weihanglo** [comment]
BTW, I added status tracking to the PR description for each backend. Let me know if anyone wants to update the status :)

**Comment by weihanglo** [comment]
**Status update**:  We have a draft PR out for unreleased libgit2 integration: <https://github.com/rust-lang/cargo/pull/16939>. Still have some snapshot cleanup but it is in a good state I believe.

**Comment by weihanglo** [comment]
**Status update:** https://crates.io/crates/git2/0.21.0 is out. Most of Cargo team members will be at RustWeek / Rust all hands this week so I doubt there will have any progress, though, though everything is possible during Rust all hands :)

The major blocker of #16939 is https://github.com/rust-lang/cargo/pull/16939#discussion_r3209423129, which we need to respect who don't want to vendor libgit2. I may try to talk to people during all hands. The brighter future is upstream libgit2 release v2 in the next few weeks or so, see <https://github.com/rust-lang/git2-rs/issues/1258> for git2-rs side issue.
